modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -318,6 +327,66 @@ public boolean authenticate(String userName, PublicKey key,
         }
     }
 
+    private class RemoteHostValidator {
+
+        private boolean isValidHost(String line, ServerSession session) {
```

Comment:
I may have missed it, but the ssh documentation doesn't really say how multiple entries should be combined to come up with an overall authorisation decision.  I had to go to the C source to get a definitive answer.  Basically any deny statements (i.e., starting with `!`) matches the the request is denied, otherwise the request is accepted only if it matches one or more allow statements (i.e., not starting with `!`).

The current semantics of this method are wrong: it short-cuts the tests if an allow pattern matches, it returns false if an allow pattern matches, and it returns true of no pattern matches.

Finally, given the rather complicated nature of parsing (and poor documentation from OpenSSH), it would be good to include some unit tests, so we can capture what "correct behaviour" means.